src/utils/ao.ts

```diff
@@ -85,3 +85,51 @@ export async function spawnANT({
 
   return processId;
 }
+
+export async function evolveANT({
+  signer,
+  processId,
+  luaCodeTxId = ANT_LUA_ID,
+  ao = connect(),
+}: {
+  signer: ContractSigner;
+  processId: string;
+  luaCodeTxId?: string;
+  ao?: AoClient;
+}): Promise<string> {
+  //TODO: cache locally and only fetch if not cached
+  const luaString = (await defaultArweave.transactions.getData(luaCodeTxId, {
+    decode: true,
+    string: true,
+  })) as string;
+
+  const aosClient = new AOProcess({
+    processId,
+    ao,
+  });
+
+  // we call eval twice to ensure the Source code tx id is set in the process. ANTs that do not have the prepended evolve handler will not do it immediately until eval is called again.
+  // TODO: evaluate gas cost of this compared to checking the with a read first vs just calling it twice.
+  // we would not need this with a weavedrive implementation of "evolve"
+  await aosClient.send({
+    tags: [
+      { name: 'Action', value: 'Eval' },
+      { name: 'App-Name', value: 'ArNS-ANT' },
+      { name: 'Source-Code-TX-ID', value: luaCodeTxId },
+    ],
+    data: luaString,
+    signer,
+  });
+
+  const { id } = await aosClient.send({
+    tags: [
+      { name: 'Action', value: 'Eval' },
+      { name: 'App-Name', value: 'ArNS-ANT' },
+      { name: 'Source-Code-TX-ID', value: luaCodeTxId },
+    ],
+    data: luaString,
+    signer,
+  });
```

Comment:
It occured to me that would prevent writing twice, but in the event that it does not have the id you wind up actually sending three network calls instead of two, so I figured it was more efficient to do it this way... now its less network calls but the second call is heavier because its sending the code twice. Gas is also something to think about here, since that info check costs gas, and im not sure what the ratio is compared to sending eval twice, or verifying with info first.

Reply:
i favor calling getInfo to avoid unnecessary evals